manager/state/raft/storage/storage.go

```diff
@@ -0,0 +1,382 @@
+		// If wals created by the etcd-v2 wal code exist, read the latest ones based
+		// on this snapshot and encode them to wals in the new path to avoid adding
+		// backwards-incompatible entries to those files.
+		legacyWALDir := e.legacyWALDir()
+		if !wal.Exist(legacyWALDir) {
+			return nil, ErrNoWAL
+		}
+
+		if err = MigrateWALs(legacyWALDir, walDir, OriginalWAL, walFactory, walsnap, log.G(ctx)); err != nil {
+			return nil, err
+		}
+	}
+
+	walObj, result.Metadata, result.HardState, result.Entries, err = ReadRepairWAL(walDir, walsnap, walFactory, log.G(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	e.snapshotter = snapshotter
+	e.wal = walObj
+
+	return result, nil
+}
+
+// BootstrapNew creates a new snapshotter and WAL writer, expecting that there is nothing on disk
+func (e *EncryptedRaftLogger) BootstrapNew(raftNode *api.RaftMember) ([]byte, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	encoder, decoder := encryption.Defaults(e.EncryptionKey)
+	walFactory := NewWALFactory(encoder, decoder)
+
+	for _, dirpath := range []string{e.walDir(), e.snapDir()} {
+		if err := os.MkdirAll(dirpath, 0700); err != nil {
+			return nil, errors.Wrap(err, "failed to create WAL directory")
+		}
+	}
+
+	metadata, err := raftNode.Marshal()
+	if err != nil {
+		return nil, errors.Wrap(err, "error marshalling raft node")
+	}
+	e.wal, err = walFactory.Create(e.walDir(), metadata)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create WAL")
+	}
+
+	e.snapshotter = NewSnapFactory(encoder, decoder).New(e.snapDir())
+	return metadata, nil
+}
+
+func (e *EncryptedRaftLogger) legacyWALDir() string {
+	return filepath.Join(e.StateDir, "wal")
+}
+
+func (e *EncryptedRaftLogger) walDir() string {
+	return filepath.Join(e.StateDir, "wal-v3")
+}
+
+func (e *EncryptedRaftLogger) legacySnapDir() string {
+	return filepath.Join(e.StateDir, "snap")
+}
+
+func (e *EncryptedRaftLogger) snapDir() string {
+	return filepath.Join(e.StateDir, "snap-v3")
+}
+
+// SaveSnapshot actually saves a given snapshot - if a new key is provided.  If a new key is provided, the key
+// is actually rotated - otherwise we always try to set the encoders to our current encryption key
+func (e *EncryptedRaftLogger) SaveSnapshot(snapshot raftpb.Snapshot, newKey []byte) error {
+	err := func() error {
+		// we want a write lock instead so we can rotate the encryption key and
+		// ensure we save the WAL snapshot with the new encryption key - this way
+		// we end up with all the WALs between 2 snapshots encrypted the same
+		e.mu.Lock()
+		defer e.mu.Unlock()
+
+		if e.wal == nil { // if the wal exists, the snapshotter exists
+			return fmt.Errorf("raft WAL has either been closed or has never been created")
+		}
+
+		walsnap := walpb.Snapshot{
+			Index: snapshot.Metadata.Index,
+			Term:  snapshot.Metadata.Term,
+		}
+
+		if newKey == nil {
+			// never mind with rotating the encoders - just save the WAL snapshot
+			return e.wal.SaveSnapshot(walsnap)
+		}
+
```

Comment:
If the key was changed, we may want to somehow force a WAL rotation. This needs more research.